src/main/scala/io/iohk/ethereum/vm/WorldStateProxy.scala

```diff
@@ -13,35 +13,39 @@ import io.iohk.ethereum.crypto.kec256
   * should be kept in memory and applied only after a transaction completes without errors. This does not forbid mutable
   * caches for DB retrieval operations.
   */
-trait WorldStateProxy {
-  type WS <: WorldStateProxy
-
-  def getAccount(address: Address): Option[Account]
-  def getCode(codeHash: ByteString): ByteString
-  def getStorage(storageRoot: ByteString): Storage
-
-  def saveAccount(address: Address, account: Account): WS
-  def saveCode(codeHash: ByteString, code: ByteString): WS
-  def saveStorage(storageRoot: ByteString, storage: Storage): WS
+trait WorldStateProxy[WS <: WorldStateProxy[WS, S], S <: Storage[S]] {
 
+  protected def getAccount(address: Address): Option[Account]
+  protected def saveAccount(address: Address, account: Account): WS
```

Comment:
This is another reason (not just boilerplate), why I wanted the opcodes to operate simply on `WorldStateProxy` and `Storage` instead of `[W <: WorldStateProxy[W, S], S <: Storage[S]]` - they don't even need to know that there are some types that implement these traits.

But that requires type-casting and I can't think of a way to do this quickly and elegantly.